File: app/src/main/java/com/storebook/inventoryapp/data/billing/BillingEngine.kt
package com.storebook.inventoryapp.data.billing

import com.storebook.inventoryapp.data.repository.CartItem
import java.math.BigDecimal
import java.math.RoundingMode
import kotlin.math.max
import kotlin.math.min

// Enum to define Tax Types
enum class TaxType {
    INTRASTATE, // CGST + SGST
    INTERSTATE, // IGST
}

// Data class to represent the calculated tax for an item
data class ItemTaxDetails(
    val cartItem: CartItem,
    val netAmountBeforeTax: Double, // Amount after discount but before tax
    val cgstAmount: Double,
    val sgstAmount: Double,
    val igstAmount: Double,
    val totalTaxAmount: Double,
    val totalAmountWithTax: Double,
)

// Data class to represent the full invoice tax calculation
data class InvoiceTaxSummary(
    val subTotal: Double,
    val totalDiscount: Double,
    val netTaxableAmount: Double, // Subtotal - Discount
    val totalCgst: Double,
    val totalSgst: Double,
    val totalIgst: Double,
    val grandTotal: Double,
    val itemDetails: List<ItemTaxDetails>,
)

object BillingEngine {
    /**
     * Extracts state code from GSTIN. First 2 characters of a valid GSTIN represent the state code.
     */
    fun getStateCodeFromGSTIN(gstin: String?): String? {
        if (gstin.isNullOrBlank() || gstin.length < 2) return null
        return gstin.substring(0, 2)
    }

    /**
     * Determines whether the transaction is Intrastate (CGST+SGST) or Interstate (IGST).
     * If both state codes match, it's Intrastate. Otherwise, Interstate.
     * If GSTIN is not provided or invalid, defaults to Intrastate for local B2C sales.
     */
    fun determineTaxType(
        businessGstin: String?,
        customerGstin: String?,
    ): TaxType {
        val bizState = getStateCodeFromGSTIN(businessGstin)
        val custState = getStateCodeFromGSTIN(customerGstin)

        return if (bizState != null && custState != null && bizState != custState) {
            TaxType.INTERSTATE
        } else {
            TaxType.INTRASTATE
        }
    }

    /**
     * Calculates tax on the NET amount (after discounts).
     * Distributes discount proportionally across items before calculating tax to avoid calculation errors.
     */
    fun calculateInvoiceTaxes(
        cartItems: List<CartItem>,
        totalDiscount: Double,
        businessGstin: String?,
        customerGstin: String?,
    ): InvoiceTaxSummary {
        val taxType = determineTaxType(businessGstin, customerGstin)

        var subTotal = BigDecimal.ZERO
        for (item in cartItems) {
            val price = BigDecimal(item.item.sellPrice.toString())
            val qty = BigDecimal(item.quantity.toString())
            subTotal = subTotal.add(price.multiply(qty))
        }

        val discount = BigDecimal(totalDiscount.toString())
        val actualDiscount = if (discount.compareTo(subTotal) > 0) subTotal else discount
        val netTaxableAmount = subTotal.subtract(actualDiscount)

        // If subtotal is 0, nothing to calculate
        if (subTotal.compareTo(BigDecimal.ZERO) <= 0) {
            return InvoiceTaxSummary(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, emptyList())
        }

        var totalCgst = BigDecimal.ZERO
        var totalSgst = BigDecimal.ZERO
        var totalIgst = BigDecimal.ZERO
        val itemDetails = mutableListOf<ItemTaxDetails>()

        val scale = 4 // Intermediate calculation precision
        val roundMode = RoundingMode.HALF_UP

        for (cartItem in cartItems) {
            val itemGross = BigDecimal(cartItem.item.sellPrice.toString()).multiply(BigDecimal(cartItem.quantity.toString()))
            val itemDiscountRatio = if (subTotal.compareTo(BigDecimal.ZERO) > 0) {
                itemGross.divide(subTotal, scale, roundMode)
            } else {
                BigDecimal.ZERO
            }
            val itemDiscount = actualDiscount.multiply(itemDiscountRatio)
            val itemNetTaxable = itemGross.subtract(itemDiscount).max(BigDecimal.ZERO)

            val taxRate = BigDecimal(cartItem.item.taxRate.toString())
            var cgst = BigDecimal.ZERO
            var sgst = BigDecimal.ZERO
            var igst = BigDecimal.ZERO

            if (taxType == TaxType.INTRASTATE) {
                val halfRate = taxRate.divide(BigDecimal("2"), scale, roundMode)
                cgst = itemNetTaxable.multiply(halfRate).divide(BigDecimal("100"), scale, roundMode)
                sgst = itemNetTaxable.multiply(halfRate).divide(BigDecimal("100"), scale, roundMode)
            } else {
                igst = itemNetTaxable.multiply(taxRate).divide(BigDecimal("100"), scale, roundMode)
            }

            val itemTotalTax = cgst.add(sgst).add(igst)

            totalCgst = totalCgst.add(cgst)
            totalSgst = totalSgst.add(sgst)
            totalIgst = totalIgst.add(igst)

            itemDetails.add(
                ItemTaxDetails(
                    cartItem = cartItem,
                    netAmountBeforeTax = itemNetTaxable.setScale(2, roundMode).toDouble(),
                    cgstAmount = cgst.setScale(2, roundMode).toDouble(),
                    sgstAmount = sgst.setScale(2, roundMode).toDouble(),
                    igstAmount = igst.setScale(2, roundMode).toDouble(),
                    totalTaxAmount = itemTotalTax.setScale(2, roundMode).toDouble(),
                    totalAmountWithTax = itemNetTaxable.add(itemTotalTax).setScale(2, roundMode).toDouble(),
                ),
            )
        }

        val grandTotal = netTaxableAmount.add(totalCgst).add(totalSgst).add(totalIgst)

        return InvoiceTaxSummary(
            subTotal = subTotal.setScale(2, roundMode).toDouble(),
            totalDiscount = actualDiscount.setScale(2, roundMode).toDouble(),
            netTaxableAmount = netTaxableAmount.setScale(2, roundMode).toDouble(),
            totalCgst = totalCgst.setScale(2, roundMode).toDouble(),
            totalSgst = totalSgst.setScale(2, roundMode).toDouble(),
            totalIgst = totalIgst.setScale(2, roundMode).toDouble(),
            grandTotal = grandTotal.setScale(2, roundMode).toDouble(),
            itemDetails = itemDetails,
        )
    }
}

File: web/src/lib/BillingEngine.ts
export enum TaxType {
  INTRASTATE = 'INTRASTATE', // CGST + SGST
  INTERSTATE = 'INTERSTATE', // IGST
}

export interface ItemTaxDetails {
  cartItemId: string;
  netAmountBeforeTax: number;
  cgstAmount: number;
  sgstAmount: number;
  igstAmount: number;
  totalTaxAmount: number;
  totalAmountWithTax: number;
}

export interface InvoiceTaxSummary {
  subTotal: number;
  totalDiscount: number;
  netTaxableAmount: number;
  totalCgst: number;
  totalSgst: number;
  totalIgst: number;
  grandTotal: number;
  itemDetails: ItemTaxDetails[];
}

export class BillingEngine {
  static getStateCodeFromGSTIN(gstin?: string | null): string | null {
    if (!gstin || gstin.trim().length < 2) return null;
    return gstin.substring(0, 2);
  }

  static determineTaxType(
    businessGstin?: string | null,
    customerGstin?: string | null
  ): TaxType {
    const bizState = this.getStateCodeFromGSTIN(businessGstin);
    const custState = this.getStateCodeFromGSTIN(customerGstin);

    if (bizState && custState && bizState !== custState) {
      return TaxType.INTERSTATE;
    }
    return TaxType.INTRASTATE;
  }

  static calculateInvoiceTaxes(
    cartItems: { id: string; sell_price: number; quantity: number; taxRate?: number }[],
    totalDiscount: number,
    businessGstin?: string | null,
    customerGstin?: string | null
  ): InvoiceTaxSummary {
    const taxType = this.determineTaxType(businessGstin, customerGstin);

    let subTotal = 0;
    for (const item of cartItems) {
      subTotal += item.sell_price * item.quantity;
    }

    const actualDiscount = Math.min(totalDiscount, subTotal);
    const netTaxableAmount = subTotal - actualDiscount;

    if (subTotal <= 0) {
      return {
        subTotal: 0,
        totalDiscount: 0,
        netTaxableAmount: 0,
        totalCgst: 0,
        totalSgst: 0,
        totalIgst: 0,
        grandTotal: 0,
        itemDetails: [],
      };
    }

    let totalCgst = 0;
    let totalSgst = 0;
    let totalIgst = 0;
    const itemDetails: ItemTaxDetails[] = [];

    for (const item of cartItems) {
      const itemGross = item.sell_price * item.quantity;
      const itemDiscountRatio = subTotal > 0 ? itemGross / subTotal : 0;
      const itemDiscount = actualDiscount * itemDiscountRatio;
      const itemNetTaxable = Math.max(0, itemGross - itemDiscount);

      const taxRate = item.taxRate || 0;
      let cgst = 0;
      let sgst = 0;
      let igst = 0;

      if (taxType === TaxType.INTRASTATE) {
        const halfRate = taxRate / 2;
        cgst = (itemNetTaxable * halfRate) / 100;
        sgst = (itemNetTaxable * halfRate) / 100;
      } else {
        igst = (itemNetTaxable * taxRate) / 100;
      }

      const itemTotalTax = cgst + sgst + igst;

      totalCgst += cgst;
      totalSgst += sgst;
      totalIgst += igst;

      itemDetails.push({
        cartItemId: item.id,
        netAmountBeforeTax: Number(itemNetTaxable.toFixed(2)),
        cgstAmount: Number(cgst.toFixed(2)),
        sgstAmount: Number(sgst.toFixed(2)),
        igstAmount: Number(igst.toFixed(2)),
        totalTaxAmount: Number(itemTotalTax.toFixed(2)),
        totalAmountWithTax: Number((itemNetTaxable + itemTotalTax).toFixed(2)),
      });
    }

    const grandTotal = netTaxableAmount + totalCgst + totalSgst + totalIgst;

    return {
      subTotal: Number(subTotal.toFixed(2)),
      totalDiscount: Number(actualDiscount.toFixed(2)),
      netTaxableAmount: Number(netTaxableAmount.toFixed(2)),
      totalCgst: Number(totalCgst.toFixed(2)),
      totalSgst: Number(totalSgst.toFixed(2)),
      totalIgst: Number(totalIgst.toFixed(2)),
      grandTotal: Number(grandTotal.toFixed(2)),
      itemDetails,
    };
  }
}
